### PR TITLE
Reduce size of sold animal lists through life of simulation

### PR DESCRIPTION
**Pull request by JoeWaddell**
### Context

Issue(s) mitigated by this pull request: #1436
Runtime is still high, this PR seems to reduce runtime by ~15%.

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
- [x] The default animal JSON has been reverted to default.

### What

Currently, the various sold animal lists simply increase in size for the life of the simulation. In this PR instead of keeping the whole animal object, we're keeping a dictionary of just the necessary outputs. 

### Why

Ideally this should help reduce runtime of RuFaS, without losing output information. 

### How

Added logic to various culling methods in `life_history.py`, changed name of lists of animals to reflect that it is a list of information, not the objects themselves.
Likewise modified the methods in `animal_module_reporter.py` to access a dict instead of objects.

### Test plan

 - [x]  Updated existing affected unit tests.

### CODE REVIEW

File: RUFAS/routines/animal/animal_module_reporter.py [hunk 1]
```diff
@@ -942,13 +927,15 @@ def report_daily_reports(cls, animal_manager, available_feeds: Dict[str, Dict[st
         AnimalModuleReporter.report_daily_ration(animal_manager, available_feeds)
         AnimalModuleReporter.report_daily_pen_total(animal_manager.simulation_day, animal_manager.all_pens)
         AnimalModuleReporter.report_305d_milk(animal_manager)
+        # AnimalModuleReporter.report_end_of_simulation(animal_manager.life_cycle_manager,
+        # animal_manager.simulation_day)
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_module_reporter.py]: comments can be removed

File: tests/test_simulation_engine.py [hunk 2]
```diff
@@ -43,11 +43,11 @@ def test_simulate(mocker: MockerFixture, start_time: int, end_time: int) -> None
...
-    patch_for_animal_module_reporter = mocker.patch(
-        "RUFAS.simulation_engine.routines.animal"
-        ".animal_module_reporter.AnimalModuleReporter"
-        ".report_end_of_simulation"
-    )
+    # patch_for_animal_module_reporter = mocker.patch(
+    #     "RUFAS.simulation_engine.routines.animal"
+    #     ".animal_module_reporter.AnimalModuleReporter"
+    #     ".report_end_of_simulation"
+    # )
```
Comment (allisterakun) [inline tests/test_simulation_engine.py]: comments

File: tests/test_simulation_engine.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def test_simulate(mocker: MockerFixture, start_time: int, end_time: int) -> None
     expected_simulation_time = end_time - start_time
     expected_log_message = f"Total simulation time is: {expected_simulation_time}"
     patch_for_output_manager.add_log.assert_called_with("total_simulation_time", expected_log_message, info_map)
-    patch_for_animal_module_reporter.assert_called_once_with(simulation_engine.animal_manager, 100)
+    # patch_for_animal_module_reporter.assert_called_once_with(simulation_engine.animal_manager, 100)
```
Comment (allisterakun) [inline tests/test_simulation_engine.py]: comments

File: RUFAS/routines/animal/life_cycle/life_cycle.py [hunk 4]
```diff
@@ -826,11 +846,31 @@ def _cull_cow(self, cow: Cow, sim_day: int | None = None) -> None:
...
-            self.sold_cows.append(cow)
+            self.sold_cows_info.append(
+                {
+                    "id": cow.id,
+                    "animal_type": Cow.__name__,
+                    "sold_at_day": cow.sold_at_day,
+                    "body_weight": cow.body_weight,
+                    "cull_reason": cow.cull_reason,
+                    "days_in_milk": cow.days_in_milk,
+                    "parity": cow.calves,
```
Comment (allisterakun) [inline RUFAS/routines/animal/life_cycle/life_cycle.py:873]: If all of the sold animal information dictionaries have the exact same structure, I suggest creating a TypedDict to enforce the structure.

**Review by allisterakun — APPROVED** [review]
Great job, Joe! 
Just make sure to remove the unused comments before merging.
Just one simple comment about possibly introducing a TypedDict to enforce the sold animal info structure.

**Review by jadamchick — APPROVED** [review]
Hi Joe, works well for me! I ran a few pilot herds on it and it generated reports smoothly with similar values to what was expected from previous runs. I didn't rigorously compare run time, but it was in line with what I expected for those herds. 

Reminder to change the default animal json back to original herd + pen sizes before you merge, please! 
& to update the changelog if it's not there already

File: RUFAS/routines/animal/life_cycle/life_cycle.py (hunk 4 above)
Comment (JoeWaddell) [reply RUFAS/routines/animal/life_cycle/life_cycle.py:873]: Great call: updated! Could you confirm the way I set it up looks OK?

**Pull request merged**